### ISSUE TITLE
Passenger on Nginx in Rails Development Mode does not reload Classes

### ISSUE THREAD

**Issue by omitter** [phusion/passenger#1921]
I tried really long to find the error, but I guess it lies in Passenger. Whenever I change a Controller Passenger delivers the old site until I restart nginx. Views are working fine. It's just the logic that is kept cached. For instance, when I let a method crash (raising some exception), Rails does show the error with the surrounding source code (at least after I restarted nginx). Then, I remove the crashing line, save the file and hit refresh in the browser. The shown source code view gets updated, but the error page still stays (even though there is no error anymore)! In other words, the source file got updated correctly and send to the client, but the internal logic got no update. 

Development happens in a Virtual Machine through Virtualbox, setup by Vagrant. Maybe there is a problem related to vboxfs?

Here is what I tried:
Set and unset several caching options in Rails, none did help.
- config.reload_classes_only_on_change
- config.cache_classes
- config.action_controller.perform_caching
- config.file_watcher

I tried also some options in nginx and passenger, but they didn't work either.
I followed the official instructions for setup, here is my config.

passenger.conf:

```
passenger_root /usr/share/ruby/vendor_ruby/phusion_passenger/locations.ini;
passenger_ruby /usr/local/rvm/wrappers/default/ruby;
passenger_instance_registry_dir /var/run/passenger-instreg;

```
rails.conf

```
server {
  listen 80;
  server_name rails.dev;
  root /srv/railsproject/public;
  passenger_enabled on;
  passenger_ruby /usr/local/rvm/wrappers/default/ruby;
  passenger_app_env development;
}
```

Rails is definitively started in development mode (checked in webconsole and database). And when I switch to puma its working fine. Adding tmp/always_restart.txt also solves the issue, but is unusably slow. 

Specs are:
Centos 7.3.1611 in VirtualBox 5.1.14 provisioned by vagrant 1.9.1
Rails 5.0.1
Ruby 2.4rc1 with RVM (1.28.0)
Nginx 1.10.2 (official passenger repo)
Passenger 5.1.2 (official passenger repo)

In case you need more info let me know. I would be very grateful towards anyone trying to solve this issue.

**Comment by CamJN** [comment]
I cannot reproduce the problem on my setup yet, the only definite difference being I didn't use vagrant to provision the VM. Potential differences are: my host is macOS 10.11.6, and my app isn't on a mounted volume.

Could you let me know what host OS you are using and what kind of volume your app is on?

**Comment by omitter** [comment]
The host OS is macOS 10.12.2 and the app is on a shared folder mounted as vboxfs. I'll try another host machine, just too be sure, but it shouldn't matter (that's point of a virtual machine, isn't it).

**Comment by omitter** [comment]
Guess my assumption was totally wrong, behaviour is indeed very host dependent. After some investigation, I found that it's all working fine for an Archlinux host.
On the initial macOS 10.12 everything works fine until the machine goes to sleep. I checked the clocks, but they are synchronized. Restarting nginx syncs the filesystem state and nginx' state, but only once. If I want to achieve the initial state (all works fine), I have to reboot the VM. I've have not yet find out how, but puma also sometimes enters this unsynced state (but far less often).
On a different mac the problem was even present without going to sleep first (nearly always that is).

I am quite overwhelmed by all these different results, but it's getting more and more unlikely that passenger is the problem. Rather the macOS virtual box implementation or some vagrant related bug. I am sorry for having troubled you.
